dashboard/src/components/api/codenvy-machine-auth.factory.js

```diff
@@ -0,0 +1,72 @@
+/*
+ *  [2015] - [2016] Codenvy, S.A.
+ *  All Rights Reserved.
+ *
+ * NOTICE:  All information contained herein is, and remains
+ * the property of Codenvy S.A. and its suppliers,
+ * if any.  The intellectual and technical concepts contained
+ * herein are proprietary to Codenvy S.A.
+ * and its suppliers and may be covered by U.S. and Foreign Patents,
+ * patents in process, and are protected by trade secret or copyright law.
+ * Dissemination of this information or reproduction of this material
+ * is strictly forbidden unless prior written permission is obtained
+ * from Codenvy S.A..
+ */
+'use strict';
+
+export class CodenvyMachineAuth {
+
+         /**
+          * Default constructor that is using resource
+          * @ngInject for Dependency injection
+          */
+         constructor($resource, $q) {
+           this.$resource = $resource;
+           this.$q = $q;
+           this.tokens = new Map();
+
+           this.tokenApi = this.$resource('/api/machine/token', {}, {
+             getByWorkspace: {method: 'GET', url: '/api/machine/token/:workspaceId'}
+           })
+         }
+
+         requestToken(workspaceId) {
+           let promise = this.tokenApi.getByWorkspace({workspaceId: workspaceId}).$promise;
+           promise.then((machineToken) => {
+             this.tokens.set(workspaceId, machineToken.machineToken);
+           });
```

Comment:
your code is used there $q.when(this.tokens.get(workspaceId) || requestToken(workspaceId))

$q.when is using promise or value
https://docs.angularjs.org/api/ng/service/$q#when

so as return, you should either return promise (but then you'll have a .machineToken inner field)
then you'll probably need to create a new request with $q.defer(); and return promise of this object
something like
https://github.com/eclipse/che/blob/ad5d3d9d7f084c8fa6441a279c42bc82d4dc4cb1/dashboard/src/components/api/che-recipe-template.factory.js#L58-61
(instead of true you return the token)